src/Test/Hspec/Wai/Servant/Assertions.hs

```diff
@@ -41,3 +42,11 @@ succeed action = do
      then getTestResponse tresp
      else do liftIO . expectationFailure $ "response error:\n  expected: status 200\n  but got: " ++ show sres
              error "unreachable"
+
+-- | Checks if the provided @action@ returns anything other than a 200. If so, sally forth.
+dontSucceed :: HasCallStack => W.WaiSession (TestResponse a) -> W.WaiSession ()
+dontSucceed action = do
+  TestResponse _ sres@(SResponse status _ _) <- action
```

Comment:
To catch decoding errors and the like, I think you should also run the function in `TestResponse` and confirm it throws if this is a `200` using [`shouldThrow`](https://hackage.haskell.org/package/hspec-expectations-0.8.2/docs/Test-Hspec-Expectations.html#v:shouldThrow):

```haskell
dontSucceed action = do
  TestResponse f sres@(SResponse status _ _) <- action
  if status /= ok200
  then f sres `shouldThrow` anyException
  else -- etc

Reply:
I updated it with correct code. Although I think we lose request logging with the `shouldThrow` method. Probably have to do the exception checking ourselves to get that to work

Reply:
Ah - yes that would be one way to do it. I like this approach. :+1:

Reply:
I think that's ok because in the case of a "successful failure" you don't really care about the request. It's only when things are not as expected (when `succeed` fails) that you would really want the request for debugging.

Reply:
Well `shouldThrow` is `IO a -> Selector b -> Expectation` and `f sres` is `SResponse -> WaiSession a` :(

Reply:
do we have something like `runWaiSession :: WaiSession a -> IO a` ?